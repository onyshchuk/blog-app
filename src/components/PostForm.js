import React from 'react';
import moment from 'moment';
import RemoveApprovalModal from './RemoveApprovalModal';

class PostForm extends React.Component {
   state = {
      id: this.props.post ? this.props.post.id : '',
      title: this.props.post ? this.props.post.title : '',
      body: this.props.post ? this.props.post.body : '',
      createdAt: this.props.post ? this.props.post.createdAt : moment().valueOf(),
      error: '',
      isOpen: false
   };
   onTitleChange = e => {
      const title = e.target.value;
      this.setState(() => ({ title }));
   };
   onBodyChange = e => {
      const body = e.target.value;
      this.setState(() => ({ body }));
   };
   onSaveClick = e => {
      e.preventDefault();

      if(this.state.title && this.state.body){
         this.props.handleSaveClick({
            id: this.state.id,
            title: this.state.title,
            body: this.state.body,
            createdAt: this.state.createdAt
         });
         this.setState({ error: '' });
      } else {
         this.setState({ error: 'Pls fill the fields' });
      }
   };
   onRemoveClick = () => {
      this.setState({ isOpen: true });
   };
   closeModal = () => {
      this.setState({ isOpen: false });
   };
   render() {
      return (
         <form className="form" onSubmit={this.onSaveClick}>
            {this.state.error && <p className="form__error">{this.state.error}</p>}
            <input
               className="text-input"
               autoFocus
               type="text"
               value={this.state.title}
               placeholder="Post title" 
               onChange={this.onTitleChange}
            />
            <textarea
               className="textarea"
               value={this.state.body}
               placeholder="Post body" 
               onChange={this.onBodyChange}
            />
            <div>
               <button className="button" type="submit">Save Post</button>
               {this.props.handleRemoveClick && <a className="button" onClick={this.onRemoveClick}>Remove Post</a>}
            </div>
            <RemoveApprovalModal
               isOpen={this.state.isOpen}
               selectedPostTitle={this.state.title}
               handleRemoveClick={this.props.handleRemoveClick}
               closeModal={this.closeModal}
            />
         </form>
      );
   };
};

export default PostForm;