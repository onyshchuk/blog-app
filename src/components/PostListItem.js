import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

export const PostListItem = ({ id, title, body, createdAt }) => (
   <Link className="list-item" to={`/edit/${id}`}>
      <div>
         <h3 className="list-item__title">{title}</h3> 
         <span className="list-item__sub-title">{body.length < 50 ? body : body.slice(0, 49) + '...'}</span> 
      </div>
      <h3 className="list-item__data">{moment(createdAt).format('MMMM Do, YYYY')}</h3>
   </Link>
);

export default PostListItem;