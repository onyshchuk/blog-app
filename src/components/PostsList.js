import React from 'react';
import { connect } from 'react-redux'; 
import PostListitem from './PostListItem';
import selectPosts from './../selectors/posts';

const PostsList = props => {
   const posts = props.posts ? props.posts : [];
   const selectedPosts = selectPosts(posts, props.filters);
   return (
      <div className="content-container">
         <div className="list-body">
            {
               selectedPosts.length === 0 ? (
                  <div className="list-item list-item--message">
                     <span>No items</span>
                  </div>
               ) : (
                  selectedPosts.map(post => <PostListitem key={post.id} id={post.id} item={post} />)
               )
            }
         </div>
      </div>
   )
}

const mapStateToProps = state => ({
   posts: state.posts,
   filters: state.filters
});

export default connect(mapStateToProps)(PostsList);