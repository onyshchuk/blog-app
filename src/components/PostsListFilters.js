import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { setTextFilter, sortByDate, sortByTitle } from '../actions/filters';

export const PostsListFilters = props => {
   const handleTextFilterChange = e => {
      props.handleTextFilterChange(e.target.value);
   };
   const handleSortByChange = e => {
      e.target.value === 'title' ? props.handleSortByTitle() : props.handleSortByDate();
   };
   return (
      <div>
         <div className="page-header">
            <div className="content-container">
               <h1 className="page-header__title">Published posts</h1>
            </div>
         </div> 
         <div className="content-container">
            <div className="input-group">
               <div className="input-group__item">
                  <input
                     className="text-input"
                     type="text"
                     placeholder="Search posts"
                     onChange={handleTextFilterChange} 
                  />
               </div>
               <div className="input-group__item">
                  <select
                     className="select"
                     value={props.filters.sortBy}
                     onChange={handleSortByChange}
                  >
                     <option value="title">title</option>
                     <option value="date">date</option>
                  </select>
               </div>
               <div className="input-group__item">
                  <Link className="button" to="/create">Add post</Link>
               </div>
            </div>
         </div>
      </div>
   );
}

const mapStateToProps = state => ({
   filters: state.filters
});

const mapDispatchToProps = dispatch => ({
   handleTextFilterChange: text => dispatch(setTextFilter(text)),
   handleSortByDate: () => dispatch(sortByDate()),
   handleSortByTitle: () => dispatch(sortByTitle())
});

export default connect(mapStateToProps, mapDispatchToProps)(PostsListFilters);