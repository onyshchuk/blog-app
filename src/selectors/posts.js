export default (posts, { text, sortBy } ) => {
   return posts.filter(post => post.title.includes(text))
               .sort((a, b) => {
                  if(sortBy === 'title') 
                     return a.title < b.title ? -1 : a.title > b.title ? 1 : 0;
                  if(sortBy === 'date') 
                     return a.createdAt < b.createdAt ? 1 : a.createdAt > b.createdAt ? -1 : 0;
               });
}