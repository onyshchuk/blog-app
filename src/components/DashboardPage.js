import React from 'react';
import PostsListFilters from './PostsListFilters';
import PostsList from './PostsList';

const DashboardPage = () => (
   <div>
      <PostsListFilters />
      <PostsList />
   </div>
);

export default DashboardPage;