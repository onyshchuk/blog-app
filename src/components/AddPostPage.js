import React from 'react';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { startAddPost } from '../actions/posts';

const AddPostPage = props => {
   const handleSaveClick = post => {
      props.handleSaveClick(post);
      props.history.push('/');
   }
   return (
      <div>
         <div className="page-header">
            <div className="content-container">
               <h1 className="page-header__title">Add Post</h1>
            </div>
         </div>
         <div className="content-container">
            <PostForm handleSaveClick={handleSaveClick} />
         </div>
      </div>
   );
}

const mapDispatchToProps = dispatch => ({
   handleSaveClick: post => dispatch(startAddPost(post))
});

export default connect(undefined, mapDispatchToProps)(AddPostPage);