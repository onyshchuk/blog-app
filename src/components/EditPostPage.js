import React from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { startEditPost, startRemovePost } from './../actions/posts';

const EditPostPage = props => {
   if(!props.post) return <Redirect to="/"/>

   const handleSaveClick = updates => {
      props.handleSaveClick(props.post.id, updates);
      props.history.push('/');
   };
   const handleRemoveClick = () => {
      props.handleRemoveClick(props.post.id);
      props.history.push('/');
   };
   return (
      <div>
         <div className="page-header">
            <div className="content-container">
               <h1 className="page-header__title">Edit Post</h1>
               <Link className="page-header__link" to={`/read/${props.post.id}`}>Post readable at .../read/{props.post.id}</Link>
            </div>
         </div>
         <div className="content-container">
            <PostForm 
               post={props.post}
               handleSaveClick={handleSaveClick} 
               handleRemoveClick={handleRemoveClick}
               readableLink={props.history.location.pathname}
            />
         </div>
      </div>
   );
}

const mapStateToProps = (state, props) => ({
   post: state.posts.find(post => post.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
   handleSaveClick: (id, updates) => dispatch(startEditPost(id, updates)),
   handleRemoveClick: id => dispatch(startRemovePost(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);