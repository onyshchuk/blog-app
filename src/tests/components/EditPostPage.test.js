import React from 'react';
import { shallow } from 'enzyme';
import { EditPostPage } from './../../components/EditPostPage';
import posts from './../fixtures/posts';

let handleSaveClick, handleRemoveClick, history, post, wrapper;

beforeEach(() => {
   post = posts[1];
   handleSaveClick = jest.fn();
   handleRemoveClick = jest.fn();
   history = { push: jest.fn() };
   wrapper = shallow(<EditPostPage
      post={post} 
      handleSaveClick={handleSaveClick}
      handleRemoveClick={handleRemoveClick}
      history={history}
   />);
});

test('should render EditPostPage properly', () => {
   expect(wrapper).toMatchSnapshot();
});

test('should handle save click', () => {
   wrapper.find('PostForm')
          .prop('handleSaveClick')({
             title: 'new title'
          });
   expect(handleSaveClick).toHaveBeenLastCalledWith(post.id, { title: 'new title' });
   expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle remove click', () => {
   wrapper.find('PostForm')
          .prop('handleRemoveClick')(post.id);
   expect(handleRemoveClick).toHaveBeenLastCalledWith(post.id);
   expect(history.push).toHaveBeenLastCalledWith('/');
});