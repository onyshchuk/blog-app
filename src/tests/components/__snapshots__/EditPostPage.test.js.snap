// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPostPage properly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Post
      </h1>
      <Link
        className="page-header__link"
        replace={false}
        to="/read/2"
      >
        Post readable at .../read/
        2
      </Link>
    </div>
  </div>
  <div
    className="content-container"
  >
    <PostForm
      handleRemoveClick={[Function]}
      handleSaveClick={[Function]}
      post={
        Object {
          "body": "second body",
          "createdAt": 0,
          "id": "2",
          "title": "second post",
        }
      }
    />
  </div>
</div>
`;
